### PR TITLE
[APM-661] Scheduler Memory Accounting

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
https://arangodb.atlassian.net/browse/APM-661
- [x] Account for work item size.
- [x] Account for worker thread stack size.

### CODE REVIEW

File: Documentation/Metrics/arangodb_scheduler_queue_memory.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: arangodb_scheduler_jobs_done_total
```
Comment (jsteemann) [inline Documentation/Metrics/arangodb_scheduler_queue_memory.yaml:1]:
```suggestion
name: arangodb_scheduler_queue_memory
```

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 2]
```diff
@@ -194,6 +194,8 @@ DECLARE_COUNTER(arangodb_scheduler_threads_started_total,
                 "Number of scheduler threads started");
 DECLARE_COUNTER(arangodb_scheduler_threads_stopped_total,
                 "Number of scheduler threads stopped");
+DECLARE_GAUGE(arangodb_scheduler_queue_memory, std::int64_t,
+              "Number of bytes allocated for tasks in the scheduler queue.");
```
Comment (jsteemann) [inline arangod/Scheduler/SupervisedScheduler.cpp:198]: nit: this metric's description ends with a dot, the ones above don't.
I don't know what is correct, but it is inconsistent like this.
```suggestion
              "Number of bytes allocated for tasks in the scheduler queue");
```

File: Documentation/Metrics/arangodb_scheduler_stack_memory.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+name: arangodb_scheduler_stack_memory
+introducedIn: "3.12.0"
+help: |
+  Approximate stack memory usage of worker threads.
+unit: byte
```
Comment (maierlars) [inline Documentation/Metrics/arangodb_scheduler_stack_memory.yaml:5]:
```suggestion
unit: bytes
```

File: Documentation/Metrics/arangodb_scheduler_stack_memory.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+type: gauge
+category: Scheduler
+complexity: simple
```
Comment (maierlars) [inline Documentation/Metrics/arangodb_scheduler_stack_memory.yaml:8]:
```suggestion
complexity: advanced
```

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 5]
```diff
@@ -606,6 +621,8 @@ void SupervisedScheduler::runSupervisor() {
       _metricsNumAwakeThreads.operator=(numAwake);
       _metricsNumWorkingThreads.operator=(numWorking);
       _metricsNumWorkerThreads.operator=(numWorkers);
+      _metricsStackMemoryWorkerThreads.operator=(
+          numWorkers* approxWorkerStackSize);
```
Comment (jsteemann) [inline arangod/Scheduler/SupervisedScheduler.cpp:625]: nit
```suggestion
          numWorkers * approxWorkerStackSize);
```

File: arangod/Scheduler/Scheduler.h [hunk 6]
```diff
@@ -240,9 +245,6 @@ class Scheduler {
   void runCronThread();
   friend class SchedulerCronThread;
 
-  // Removed all tasks from the priority queue and cancels them
-  void cancelAllCronTasks();
```
Comment (jsteemann) [inline arangod/Scheduler/Scheduler.h:244]: :+1:

File: Documentation/Metrics/arangodb_scheduler_stack_memory.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+type: gauge
+category: Scheduler
+complexity: advanced
+exposedBy:
+  - coordinator
+  - dbserver
+  - agent
+  - single
+description: |
+  Approximate stack memory usage of worker threads.
```
Comment (jsteemann) [inline Documentation/Metrics/arangodb_scheduler_stack_memory.yaml:15]:
```suggestion
  Approximate stack memory usage of worker threads.
  Currently a hard-coded default value of 4MB is assumed per worker thread.
```

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 8]
```diff
@@ -178,6 +178,8 @@ DECLARE_GAUGE(arangodb_scheduler_num_working_threads, uint64_t,
               "Number of working threads");
 DECLARE_GAUGE(arangodb_scheduler_num_worker_threads, uint64_t,
               "Number of worker threads");
+DECLARE_GAUGE(arangodb_scheduler_stack_memory, uint64_t,
+              "Approx. stack memory usage of worker threads");
```
Comment (jsteemann) [inline arangod/Scheduler/SupervisedScheduler.cpp:182]:
```suggestion
              "Approximate stack memory usage of worker threads");
```

**Review by jsteemann — APPROVED** [review]
LGTM

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**